Bug: op failed to verify that width of result type has to be smaller than or equal to the input type

```systemverilog

typedef struct packed {
    logic                   element;    
} mystruct_t;

module my_module

#(
    parameter DEPTH = 16              
)
(
    input mystruct_t D_flopped
);

mystruct_t [1:0][DEPTH-1:0] Qall;

always_latch begin
    for ( int array=0 ; array<2 ; array++ ) begin
        for ( int row=0 ; row<DEPTH ; row++ ) begin 
                Qall[array][row] = D_flopped;
        end
    end
end 

endmodule
```

```
circt-verilog my_file.sv;
```
```
my_file.sv:20:17: error: 'llhd.sig.array_slice' op failed to verify that width of result type has to be smaller than or equal to the input type
                Qall[array][row] = D_flopped;
                ^
my_file.sv:20:17: note: see current operation: %23 = "llhd.sig.array_slice"(%9, %22) : (!llhd.ref<!hw.array<2xarray<16xstruct<element: i1>>>>, i1) -> !llhd.ref<!hw.array<16xstruct<element: i1>>>
```

This works in all big simulators. I tried forcing casting and bit concatenation but that wouldn't work either. Problem is with the struct. If i replace the struct with just logic type it runs fine.

--- a/lib/Conversion/MooreToCore/MooreToCore.cpp
+++ b/lib/Conversion/MooreToCore/MooreToCore.cpp
@@ -1214,13 +1214,15 @@ struct DynExtractOpConversion : public OpConversionPattern<DynExtractOp> {
       Value idx = adjustIntegerWidth(rewriter, adaptor.getLowBit(), idxWidth,
                                      op->getLoc());
 
-      if (isa<hw::ArrayType>(resultType)) {
+      bool isSingleElementExtract = arrType.getElementType() == resultType;
+
+      if (isSingleElementExtract)
+        rewriter.replaceOpWithNewOp<hw::ArrayGetOp>(op, adaptor.getInput(),
+                                                    idx);
+      else
         rewriter.replaceOpWithNewOp<hw::ArraySliceOp>(op, resultType,
                                                       adaptor.getInput(), idx);
-        return success();
-      }
 
-      rewriter.replaceOpWithNewOp<hw::ArrayGetOp>(op, adaptor.getInput(), idx);
       return success();
     }
 
@@ -1257,14 +1259,17 @@ struct DynExtractRefOpConversion : public OpConversionPattern<DynExtractRefOp> {
           rewriter, adaptor.getLowBit(),
           llvm::Log2_64_Ceil(arrType.getNumElements()), op->getLoc());
 
-      if (isa<hw::ArrayType>(cast<llhd::RefType>(resultType).getNestedType())) {
+      auto resultNestedType = cast<llhd::RefType>(resultType).getNestedType();
+      bool isSingleElementExtract =
+          arrType.getElementType() == resultNestedType;
+
+      if (isSingleElementExtract)
+        rewriter.replaceOpWithNewOp<llhd::SigArrayGetOp>(op, adaptor.getInput(),
+                                                         idx);
+      else
         rewriter.replaceOpWithNewOp<llhd::SigArraySliceOp>(
             op, resultType, adaptor.getInput(), idx);
-        return success();
-      }
 
-      rewriter.replaceOpWithNewOp<llhd::SigArrayGetOp>(op, adaptor.getInput(),
-                                                       idx);
       return success();
     }
 

--- a/test/Conversion/MooreToCore/basic.mlir
+++ b/test/Conversion/MooreToCore/basic.mlir
@@ -345,6 +345,20 @@ func.func @ExtractRefArrayElement(%j: !moore.ref<array<1 x array<1 x l3>>>) -> (
   return %0 : !moore.ref<array<1 x l3>>
 }
 
+// CHECK-LABEL: DynExtractArrayElement
+func.func @DynExtractArrayElement(%j: !moore.array<2 x array<1 x l3>>, %idx: !moore.l1) -> (!moore.array<1 x l3>) {
+  // CHECK: hw.array_get
+  %0 = moore.dyn_extract %j from %idx : !moore.array<2 x array<1 x l3>>, !moore.l1 -> !moore.array<1 x l3>
+  return %0 : !moore.array<1 x l3>
+}
+
+// CHECK-LABEL: DynExtractRefArrayElement
+func.func @DynExtractRefArrayElement(%j: !moore.ref<array<2 x array<1 x l3>>>, %idx: !moore.l1) -> (!moore.ref<array<1 x l3>>) {
+  // CHECK: llhd.sig.array_get
+  %0 = moore.dyn_extract_ref %j from %idx : <array<2 x array<1 x l3>>>, !moore.l1 -> <array<1 x l3>>
+  return %0 : !moore.ref<array<1 x l3>>
+}
+
 // CHECK-LABEL: func @AdvancedConversion
 func.func @AdvancedConversion(%arg0: !moore.array<5 x struct<{exp_bits: i32, man_bits: i32}>>) -> (!moore.array<5 x struct<{exp_bits: i32, man_bits: i32}>>, !moore.i320) {
   // CHECK: [[V0:%.+]] = hw.constant 3978585893941511189997889893581765703992223160870725712510875979948892565035285336817671 : i320

